packet/src/pacer.rs

```diff
@@ -0,0 +1,131 @@
+use std::time::{Duration, Instant};
+
+pub struct LeakyBucketPacer {
+    /// Target bitrate in bits per second.
+    target_bitrate: u64,
+    /// The last time we refreshed the budget.
+    last_budget_increase: Option<Instant>,
+    /// The current budget in bytes.
+    budget: usize,
+    /// The current pacing i.e. how frequently we refresh the budget.
+    pacing: Duration,
+}
+
+pub enum Decision {
+    Emit,
+    Hold,
+}
+
+impl LeakyBucketPacer {
+    pub fn set_target_bitrate(&mut self, new_rate: u64) {
+        self.target_bitrate = new_rate;
+    }
+
+    pub fn next_interval(&self) -> Option<Instant> {
+        self.last_budget_increase.map(|l| l + self.pacing)
+    }
+
+    pub fn decide(&mut self, data: &[u8], now: Instant) -> Decision {
+        self.maybe_update_budget(now);
```

Comment:
I'm thinking that rather than implicitly driving time, maybe the component should split the operation into two.

1.`handle_time(&mut self, Instant)`
2.`decide(&mut self, len: usize)`
3. `poll_timeout(&mut self) -> Option<Instant>`

This ties into how the rest of str0m operates, i.e. that driving time forward is an explicit action, while polling for output doesn't drive time.

The `now` that is available in poll_packet is more about registering send times than actually _driving_ time. At least that was the original idea.

Might be splitting hairs. What do you think?

Reply:
I was thinking the same thing: that this should be like the other "Sans I/O" things with handle_timeout and poll_timeout and poll_transmission and that kind of thing.

Reply:
Sounds like we agree. I will split this